### PR TITLE
Multi-chunk uploader

### PR DESCRIPTION
**Pull request by mudler**
See: https://progress.opensuse.org/issues/27955, https://progress.opensuse.org/issues/17788

http://e122.suse.de/tests/6309/file/worker-log.txt

~~Still needs updating apparmor rules~~ Seems no action is required for apparmor http://e122.suse.de/tests/7202

### CODE REVIEW

File: lib/OpenQA/Worker/Jobs.pm [hunk 1]
```diff
@@ -182,15 +183,37 @@ sub _reset_state {
...
     my ($job_id, $form) = @_;
-    unless (verify_workerid) {
-        _reset_state;
-        die 'No current_host!';
-    }
     my $filename = $form->{file}->{filename};
     my $file     = $form->{file}->{file};
+    my $is_asset = $form->{asset};
 
+    my $pieces = OpenQA::File->new(file => Mojo::File->new($file))->split(1024);
```
Comment (foursixnine) [inline lib/OpenQA/Worker/Jobs.pm]: This should be configurable somewhere

File: lib/OpenQA/Worker/Jobs.pm [hunk 2]
```diff
@@ -249,35 +271,23 @@ sub upload {
...
-    if ($res->res->json && $res->res->json->{temporary}) {
-        my $csum1 = '1';
-        my $size1;
-        if (open(my $cfd, "-|", "cksum", $res->res->json->{temporary})) {
-            ($csum1, $size1) = split(/ /, <$cfd>);
-            close($cfd);
-        }
-        my $csum2 = '2';
-        my $size2;
-        if (open(my $cfd, "-|", "cksum", $file)) {
```
Comment (foursixnine) [inline lib/OpenQA/Worker/Jobs.pm]: \o/ no more NFS!

File: lib/OpenQA/Worker/Jobs.pm (hunk 1 above)
Comment (mudler) [reply lib/OpenQA/Worker/Jobs.pm]: Yeah 👍

File: lib/OpenQA/Worker/Jobs.pm (hunk 2 above)
Comment (mudler) [reply lib/OpenQA/Worker/Jobs.pm]: \o/ 🎉 🎉 🎉 🎉

Comment (coolo) [reply lib/OpenQA/Worker/Jobs.pm]: careful. we have another use case for NFS: tests reference other/ assets not downloaded into cache.

**Comment by coolo** [comment]
ready?

**Comment by mudler** [comment]
It is actually, i'm just trying to optimize it to avoid to be action-blocking on the WebUI side (while computing cksum and moving the final asset )

**Comment by mudler** [comment]
Should be done by today :)

**Comment by mudler** [comment]
code inside Mojo::IOLoop::Subprocess doesn't get into codecov reports ;(

**Comment by coolo** [comment]
most likely the same problem we faced in isotovideo - where we had to add explicit
```Devel::Cover::report()``` callls in subprocesses

**Comment by mudler** [comment]
yeah, i'm trying to workaround  with ```Mojo::IOLoop->singleton->on('chunk_upload.start' => sub { Devel::Cover::report() if Devel::Cover->can('report'); } )``` and firing some event inside the subprocess - not much luck so far

**Comment by mudler** [comment]
Ok, seems good enough to go.

Just my personal opinion: the chunk size default it's 1MB, and i believe it's too small. 
Overall it's noticeable slowdown in the upload of the image, it takes almost ~1.30h to upload a file of 1.7GB, and can be seen in the worker logs (it's showing the average speed as well, e.g. [here](http://e122.suse.de/tests/7172/file/worker-log.txt)), but for sake of stability maybe it is fine.

I would rather opt for setting up in the workers that have a high latency a small chunk size, but not as default.  It should be enough anyway to set `UPLOAD_CHUNK_SIZE` inside workers.ini.

Besides that, [here](https://github.com/os-autoinst/openQA/pull/1564/files#diff-84ba0fe547a8b3d99e504df740605e4cR34) i'm trying to setup a default schema for the url if not defined in the workers.ini. I recommend that the workers's should specify the scheme instead of guessing and using http as fallback.

Tested on: http://e122.suse.de/tests/

Parallel tests publishing HDD:

http://e122.suse.de/tests/7166
http://e122.suse.de/tests/7167
http://e122.suse.de/tests/7168
http://e122.suse.de/tests/7169
http://e122.suse.de/tests/7170
http://e122.suse.de/tests/7171
http://e122.suse.de/tests/7172
http://e122.suse.de/tests/7173
http://e122.suse.de/tests/7174
http://e122.suse.de/tests/7175
http://e122.suse.de/tests/7176
http://e122.suse.de/tests/7177
http://e122.suse.de/tests/7178
http://e122.suse.de/tests/7179

**Comment by coolo** [comment]
damn, I screwed it - please resolve the conflict (by adapting the POD)

**Comment by mudler** [comment]
no problemo

**Pull request merged**